java-backend/src/main/java/org/kframework/backend/java/kil/Term.java

```diff
@@ -82,6 +81,26 @@ public Term substituteAndEvaluate(Map<Variable, ? extends Term> substitution, Te
                this;
     }
 
+    /**
+     * Returns a new {@code Term} instance obtained from this term by applying
+     * substitution from {@code context.getTopConstraint().substitution()}.
+     * <p>
+     * Evaluates all pending functions and predicates.
+     * <p>
+     * Denis, K branch `gnosis` only: Now that function evaluation results are cached,
+     * I suspect this function can be safely used instead of evaluate() and substituteAndEvaluate() in all contexts.
+     * These two no longer have performance advantages.
+     */
+    public Term fullSubstituteAndEvaluate(TermContext context) {
+        Map<Variable, ? extends Term> substitution = context.getTopConstraint().substitution();
+        Term newTerm = !Collections.disjoint(variableSet(), substitution.keySet())
+                ? (Term) this.accept(new SubstituteAndEvaluateTransformer(substitution, context))
+                : this;
```

Comment:
hmm, so, the `evaluate` function does NOT consider the `context.getTopConstraint().substitution()`?

Reply:
Yes, `evaluate()` does not perform substitution. I believe we can safely replace all instances of `evaluate()` with function above, but there's no need to do this.

Reply:
I didn't know that, and cannot see why.  There might be some specific reason why @andreistefanescu did that way.  Yes, let's do the drop-in replacement when needed later, not now.